runtime/src/main/java/com/dylibso/chicory/runtime/Memory.java

```diff
@@ -117,12 +118,40 @@ public void initPassiveSegment(int segmentId, int dest, int offset, int size) {
         write(dest, segment.data(), offset, size);
     }
 
+    public void writeString(int offset, String data, Charset charSet) {
+        write(offset, data.getBytes(charSet));
+    }
+
     public void writeString(int offset, String data) {
-        write(offset, data.getBytes(StandardCharsets.UTF_8));
+        writeString(offset, data, StandardCharsets.UTF_8);
     }
 
     public String readString(int addr, int len) {
-        return new String(readBytes(addr, len), StandardCharsets.UTF_8);
+        return readString(addr, len, StandardCharsets.UTF_8);
+    }
+
+    public String readString(int addr, int len, Charset charSet) {
+        return new String(readBytes(addr, len), charSet);
+    }
+
+    public void writeCString(int offset, String str) {
+        writeCString(offset, str, StandardCharsets.UTF_8);
+    }
+
+    public void writeCString(int offset, String str, Charset charSet) {
+        writeString(offset, str + '\0', charSet);
+    }
+
+    public String readCString(int addr, Charset charSet) {
+        int c = addr;
+        while (read(c) != '\0') {
+            c++;
```

Comment:
perhaps a better way to do this? not sure whether it's more efficient to also be reading in the string in this loop or only get the count.